### PR TITLE
Disable non-legacy themes selection on Settings screen when AMP slug is defined late

### LINKED ISSUE #5789: Prevent non-legacy themes from being selected as Reader theme on Settings screen when AMP slug defined late
## Bug Description

In #4999 we check if the AMP slug (query var) is defined too late and if so, prevent selection of a non-legacy Reader theme:

> image

This, however, was not likewise implemented for the Settings screen which allows all to be selected:

> image

## Expected Behaviour

The available themes should be consistent between the Settings screen and the Onboarding Wizard.

## Steps to reproduce

Activate this plugin:

```php
<?php
/**
 * Plugin Name: AMP Define Query Var Early with Constant
 */

add_action(
	'after_setup_theme',
	function () {
		add_filter(
			'amp_query_var',
			function() {
				return 'too-early-with-constant';
			}
		);
	},
	~PHP_INT_MAX
);
```

Then access the AMP Settings screen and select Reader template mode.

This is part of #4795.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by delawski**
## Summary

Fixes #5789

Currently, a theme that cannot be selected as a Reader theme (i.e. defines the AMP slug too late) is disabled in the Onboarding Wizard. However, we don't do the same on the AMP Settings page.

Onboarding Wizard | Settings page
--- | ---
Onboarding Wizard | Settings page

With this PR, the experience is consistent in both plugin areas:

Onboarding Wizard | Settings page
--- | ---
Screenshot 2021-02-05 at 13 38 35 | Screenshot 2021-02-05 at 13 38 12

Moreover, if there is just one page in the carousel on the Settings page, no navigation will be shown:

Before | After
--- | ---
Screenshot 2021-02-05 at 13 42 18 | Screenshot 2021-02-05 at 13 41 22

## Details

The themes are now categorized by their availability for being used as a Reader Theme in the `ReaderThemes` context provider. We then use the `availableThemes` and `unavailableThemes` lists in both `ReaderThemeCarousel` and `ReaderThemeSelection`.

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by delawski** [comment]
> Moreover, if there is just one page in the carousel on the Settings page, no navigation will be shown:

@jwold Can you check out the attached screenshots and tell if the update looks good to you?

**Comment by jwold** [comment]
@delawski thanks for sharing! The screenshots were helpful, makes it easier to get a sense of what's going on.

Here are two small suggestions with the current design:

1. The yellow [notice](https://d.pr/i/qNwx1p) may be a bit strong, instead perhaps a gray notice makes more sense (more in line with the one in the onboarding wizard
2. Instead of a [checkmark](https://d.pr/i/JKRML2) for the `show unavailable themes`, I wonder if we'd want to use a toggle (like we have in other parts of the onboarding wizard. 

And here is a larger suggestion we could look at:

If we are going to have all of the unavailable themes at the end of the list, we could play with a concept to put the notice between the available and unavailable themes. Here's an idea of how that could look:

My Sketches 3 - 2021-02-05 12 31 52

Is that feasible? What do you think?

File: assets/src/components/carousel/index.js [hunk 1]
```diff
@@ -124,7 +125,7 @@ export function Carousel( {
 					) ) }
 				</ul>
 			</div>
-			{ currentPage && (
+			{ showCarouselNav && (
```
Comment (pierlon) [inline assets/src/components/carousel/index.js:135]: I noticed the carousel navigation is not updated correctly when viewed on mobile:


https://user-images.githubusercontent.com/16200219/107400384-aace9b80-6af9-11eb-8698-8388a3c63977.mp4


A workaround may be to wrap `showCarouselNav` in a `useEffect` hook:

```diff
diff --git a/assets/src/components/carousel/index.js b/assets/src/components/carousel/index.js
--- a/assets/src/components/carousel/index.js	(revision f14aa035c6830468fb1059924477ce6688300a4d)
+++ b/assets/src/components/carousel/index.js	(date 1612887747254)
@@ -38,6 +38,7 @@
 	const { windowWidth } = useWindowWidth();
 	const [ currentPage, originalSetCurrentPage ] = useState( null );
 	const [ pageWidth, setPageWidth ] = useState( 0 );
+	const [ showCarouselNav, setShowCarouselNav ] = useState( null );
 	const carouselListRef = useRef();
 
 	/**
@@ -103,7 +104,13 @@
 		setPageWidth( carouselListRef?.current?.clientWidth || 0 );
 	}, [ items.length, windowWidth ] );
 
-	const showCarouselNav = currentPage && items.length > 1;
+	/**
+	 * Show the carousel navigation when there is more than one page.
+	 */
+	useEffect( () => {
+		setShowCarouselNav( items.length > 1 );
+	}, [ items.length ] );
+
 	const centeredItemIndex = [ ...( carouselListRef.current?.children || [] ) ].indexOf( currentPage );
 	const nextButtonDisabled = centeredItemIndex >= items.length - 1;
 	const prevButtonDisabled = centeredItemIndex <= 0;
```

Comment (delawski) [reply assets/src/components/carousel/index.js:135]: Thank you, @pierlon! I've addressed it in 5934dd8.

**Comment by delawski** [comment]
@jwold I've addressed your feedback in f172507. The `info` (light blue) notice type is now used and the checkbox is replaced with a toggle:

Screenshot 2021-02-11 at 19 29 06

I'm wondering if we should update the wording in the message, though. Right now it may be confusing since not *all* themes listed below are unavailable, just some of them.

Maybe it should say something like:

> Some of the themes below are unavailable...

instead of:

> The following themes are unavailable...

**Comment by jwold** [comment]
> I'm wondering if we should update the wording in the message, though.

Good catch @delawski . I think that wording change is better.

**Comment by delawski** [comment]
> Good catch @delawski . I think that wording change is better.

@jwold This has been updated in 6be58ae

File: assets/src/components/reader-theme-selection/index.js [hunk 2]
```diff
@@ -53,6 +37,19 @@ export function ReaderThemeSelection() {
...
+						{
+							sprintf(
+								/* translators: placeholder is the name of a WordPress theme. */
+								__( 'Your active theme “%s” is not available as a reader theme. If you wish to use it, Transitional mode may be the best option for you.', 'amp' ),
+								currentTheme.name,
+							)
+						}
+					</p>
+				</AMPNotice>
+			) }
```
Comment (pierlon) [inline assets/src/components/reader-theme-selection/index.js]: Was this change meant to be apart of the PR? I see it was included in the most recent merge commit: https://github.com/ampproject/amp-wp/pull/5859/commits/4afb475c747cc1ce27b197c88071641d4bbed6b4.

Comment (pierlon) [reply assets/src/components/reader-theme-selection/index.js]: Also, should the current theme be disabled here in the Onboarding Wizard? I remember @westonruter reverting the change (see https://github.com/ampproject/amp-wp/pull/5899#discussion_r583289143).

**Review by pierlon — CHANGES_REQUESTED** [review]

File: assets/src/components/reader-theme-selection/index.js (hunk 2 above)
Comment (delawski) [reply assets/src/components/reader-theme-selection/index.js]: Thanks, @pierlon. I think this code slipped through my fingers when I've been manually resolving merge conflicts.

**Review by pierlon — APPROVED** [review]
This looks ready to :ship:, great work @delawski!

**Review by westonruter — APPROVED** [review]
Thank you!

**Comment by jwold** [comment]
@delawski we had a chat about this on our Tuesday call:

* PG > Haven’t reviewed in a while, will review and - I don’t think there’s much that has changed. Will give my update later today.
* WR > Have idea about this as well, where - reason. On WP.com VIP everything is done in classic VIP in context of a theme; plugins put inside of theme, everything put inside theme. Nothing runs until theme is loaded, and the way you configured the AMP query bar in URL was also done in the theme; you defined AMP query bar constant in theme. If you wanted light or stripped down or ?strippdown, then you put that in your theme. But reader themes need to be running earlier than theme loads, need to switch theme entirely; needs to be done early in WP execution flow, that’s why we disable reader themes when you have defined slug too late. However, in working on the issue with the editor or theme colors, editor color palette; I realized that we could rather store slug in database as well so that even though you defined slug in code, we could essentially cache that so that we can use it earlier in the WP execution cycle even though it’s defined later. 
* WR > Way around this problem
* WR > Number of sites defining slug this late probably low, probably not important to worry about. 
* WR > [As far as next actions] maybe we comment on the issue with the notes, and then think about it for future. 

---

**Pull request merged**